provider/datagen/src/lib.rs

```diff
@@ -2,288 +2,219 @@
-                $crate::cldr::PluralsProvider,
-                $crate::cldr::TimeZonesProvider,
-                $crate::cldr::WeekDataProvider,
-                $crate::cldr::ListProvider,
-                $crate::uprops::CaseMappingDataProvider,
-                $crate::uprops::EnumeratedPropertyCodePointTrieProvider,
-                $crate::uprops::ScriptWithExtensionsPropertyProvider,
-                $crate::uprops::EnumeratedPropertyUnicodeSetDataProvider,
-                // Has to go last as it matches all props/ keys.
-                $crate::uprops::BinaryPropertyUnicodeSetDataProvider,
-                $crate::segmenter::SegmenterRuleProvider,
-            ]
-        )
-    };
-    ($source_data:expr, [ $($constructor:path),+, ]) => {{
-        let __source = $source_data;
-        icu_provider_adapters::make_forking_provider!(
-            icu_provider_adapters::fork::by_key::ForkByKeyProvider,
-            [
-                icu_provider::hello_world::HelloWorldProvider::new_with_placeholder_data(),
+pub fn keys_from_file<R: Read>(file: R) -> std::io::Result<Vec<ResourceKey>> {
+    Ok(keys(&BufReader::new(file).lines().collect::<std::io::Result<Vec<String>>>()?))
+}
 
-                $(<$constructor>::from(&__source)),+,
-            ]
-        )
-    }};
+/// The output format.
+#[non_exhaustive]
+pub enum Out {
+    /// Output to a file system tree
+    Fs {
+        /// The root path.
+        output_path: PathBuf,
+        /// The serialization format. See [icu_provider_fs::export::serializers].
+        serializer: Box<dyn serializers::AbstractSerializer + Sync>,
+        /// Whether to overwrite existing data.
+        overwrite: bool,
+    },
+    /// Output as a postcard blob to the given sink.
+    Blob(Box<dyn std::io::Write + Sync>),
 }
 
-#[test]
-fn no_key_collisions() {
-    let mut map = std::collections::BTreeMap::new();
-    let mut failed = false;
-    for key in get_all_keys() {
-        if let Some(colliding_key) = map.insert(key.get_hash(), key) {
-            println!(
-                "{:?} and {:?} collide at {:?}",
-                key.get_path(),
-                colliding_key.get_path(),
-                key.get_hash()
-            );
-            failed = true;
-        }
-    }
-    if failed {
-        panic!();
+/// Runs ICU4X datagen.
+///
+/// The argument are used as follows:
+/// * `locales`: If this is present, only locales that are either `und` or
+///   contained (strictly, i.e. `en` != `en-US`) in the slice will be generated.
+///   Otherwise, all locales supported by the source data will be generated.
+/// * `keys`: The keys for which to generate data. See [`get_all_keys()`].
+/// * `sources`: The underlying source data. CLDR and/or uprops data can be missing if no
+///   requested key requires them. Otherwise a error will be returned that can be identified
+///   with [`is_missing_cldr_error`] or [`is_missing_uprops_error`].
+/// * `out`: See the documentation on [`Out`]
```

Comment:
nit: should probably still say something like "details on the output format and location. see documentation on Out"